run_matrix.py

```diff
@@ -0,0 +1,110 @@
+
+
+def _clean(root: Path = Path("/tmp/charm-relation-interfaces-tests/")):
+    if root.is_dir():
+        shutil.rmtree(root)
+
+
+def _generate_test(interface: str, test_path: Path, fixture_id: str) -> Path:
+    test_content = f"""from interface_tester import InterfaceTester
+    
+def test_{interface}_interface({fixture_id}: InterfaceTester):
+    {fixture_id}.configure(
+        interface_name="{interface}",
+    )
+    {fixture_id}.run()
+    """
+    test_filename = f"interface-test-{interface}.py"
+    with open(test_path / test_filename, "w") as file:
+        file.write(test_content)
+    return test_path / test_filename
+        
+
+def _get_fixture(charm_config: _CharmTestConfig, charm_path: Path):
+    fixture_path = charm_path / FIXTURE_PATH
+    fixture_id = FIXTURE_IDENTIFIER
+    if charm_config.test_setup:
+        if charm_config.test_setup["location"]:
+            fixture_path = charm_path / Path(charm_config.test_setup["location"])
+        if charm_config.test_setup["identifier"]:
+            fixture_id = charm_config.test_setup["identifier"]
+    return fixture_path, fixture_id
+
+
+def _prepare_venv(charm_path: Path) -> Path:
+    """Create the venv for a charm and return the path to its python."""
+    logging.info(f"Installing dependencies in venv for {charm_path}")
+    original_wd = os.getcwd()
+    os.chdir(charm_path)
+    # Create the venv and install the requirements
+    os.system("python -m venv ./.interface-venv >/dev/null")
+    os.system(".interface-venv/bin/python -m pip install setuptools pytest git+https://github.com/PietroPasotti/interface-tester-pytest@main >/dev/null 2>&1")
+    os.system(".interface-venv/bin/python -m pip install -r requirements.txt >/dev/null 2>&1")
+    os.chdir(original_wd)
+    return charm_path / ".interface-venv/bin/python"
+
+
+def test_charm(charm_path: Path, test_path: Path):
+    logging.info(f"Running tests for {charm_path}")
+    original_wd = os.getcwd()
+    os.chdir(charm_path)
+    os.system(f"PYTHONPATH=src:lib .interface-venv/bin/python -m pytest {test_path}")
+    os.chdir(original_wd) 
+
+
+def run_interface_tests(path: Path, include: str = "*"):
+    test_results = {}
+    for interface, x in collect_tests(path=path, include=include).items():
+        test_results[interface] = {}
+        logging.info(f"Running tests for interface: {interface}")
+        for _, y in x.items():
+            for role in ["provider", "requirer"]:
+                test_results[interface][role] = {}
+                for charm_config in y[role]["charms"]:
+                    last_result = "success"
+                    try:
+                        logging.info(f"Charm: {charm_config.name}")   
+                        charm_path, test_path = prepare_repo(charm_config, interface)
+                        test_charm(charm_path, test_path)
+                        logging.info(f"Finished tests for charm {charm_config.name}")
+                    except Exception as e:
```

Comment:
I would refactor `prepare_repo` to raise a SetupError (or something like that) and `test_charm` to raise an InterfaceTestError (or something like that), and catch them here separately. Which part of the test failed is valuable information: it tells us where to file the issue.